An isocrystal story (#47)
leanprover-community/blog: posts/formalizing-isocrystals-in-lean.md

---
author: 'Robert Y. Lewis, Heather Macbeth'
category: 'New in mathlib'
date: 2022-07-19 19:30:08 UTC+02:00
description: ''
has_math: true
link: ''
slug: classification-of-one-dimensional-isocrystals
tags: ''
title: Classification of one-dimensional isocrystals
type: text
---

> Last year, there was a [big mathlib refactor](https://leanprover-community.github.io/blog/posts/semilinear-maps/) to replace linear maps throughout the library with  *semilinear maps*,

Reviewer: ```suggestion
Last year, there was a [big mathlib refactor](link://filename/semilinear-maps.md) to replace linear maps throughout the library with  *semilinear maps*,
```
which ensures that the links don't break if we rename the md (which I assume would give a build error) or move the blog root url.

Author: Local testing implies it needs the full path:
```suggestion
Last year, there was a [big mathlib refactor](link://filename/posts/semilinear-maps.md) to replace linear maps throughout the library with  *semilinear maps*,
```

a more abstract concept which, importantly, unifies linear and conjugate-linear maps.

But this is not the full extent of the generalization!  Our number theorist friends here in mathlib told us that we should
make sure we chose this full generality of *semilinear* maps, maps $f:M \to N$ such that $f(ax)=\sigma(a)f(x)$ for some ring homomorphism $\sigma$ between the scalar rings of the modules $M$ and $N$.  So we and our coauthor Frédéric Dupuis implemented this full generality, and then asked them for an example to illustrate their need for this more abstract definition. This blog post tells the story of our little adventure in number theory.

> It turns out that the standard use of semilinear maps in number theory is for *Frobenius-semilinearity*, semilinearity with
> respect to the ring homomorphism of the fraction field of the $p$-typical Witt vectors over a perfect characteristic-$p$ integral domain which is induced by the Frobenius
> automorphism of that domain.  Let's backtrack to catch everyone up...

Reviewer 2: Since this is the "teaser", you could consider saying something like
```suggestion
It turns out that the standard use of semilinear maps in number theory is for *Frobenius-semilinearity*,
semilinearity with respect to ring homomorphisms that are related to the Frobenius homomorphisms $x ↦ x^p$
that show up in the study of finite fields of characteristic $p$.
In our work, this Frobenius homomorphism manifested itself as an automorphism of
the fraction field of the $p$-typical Witt vectors over a perfect characteristic-$p$ integral domain
which is induced by the Frobenius automorphism of that domain.
Let's backtrack to catch everyone up...
```
That way the first reminder on Frobenius is maybe a little bit less intimidating.

Reviewer 3: Why don't you even prefer "study of algebras over fields of characteristic $p$"? There is no good reason to stick to finite ones, and in particular most of the geometric instances come from infinite F_p-algebras, no?

Reviewer 4: If you want to make it even less intimidating, you could simply write "It turns out that one standard use of semilinear maps in number theory is *Frobenius-semilinearity*, semilinearity with respect to the Frobenius endomorphism associated to a prime number p. The Frobenius endomorphism can be viewed as a non-archimedean analogue of complex conjugation, so perhaps one could argue that the two different motivations for formalising semilinear maps boil down to an archimedean and a nonarchimedean avatar of the same underlying idea.

Reviewer 3: This is also a very nice way of putting it, yes.

Author: I think the stylistic point of this sentence before the teaser was to give a *very* small taste of what's coming, postponing the real details and exposition for later. It's supposed to be intimidating for effect. (Hence "let's backtrack...") So I'm not sure expanding it (beyond making sure it's technically correct) is the right move. Heather, what do you think?

<!-- TEASER_END -->

# Witt vectors and isocrystals

Back in 2020, one of us (Rob), together with Johan Commelin, [formalized the theory of Witt vectors](https://arxiv.org/abs/2010.02595) in Lean.  The $p$-typical Witt 
vectors $\mathbb{W}(R)$ over a ring $R$  are, concretely, sequences of elements of $R$, but equipped with a rather wild multiplicative and additive

> structure (dependent on $p$) to make this set into a commutative ring.  The canonical example is the $p$-adic numbers, which are the Witt vectors

Reviewer 4: ```suggestion
structure (dependent on $p$) to make this set into a commutative ring.  The canonical example is the $p$-adic integers, which are the Witt vectors
```

of $\mathbb{Z}/p\mathbb{Z}$.

A ring $R$ of characteristic $p$ has an endomorphism, constructed by sending each element $x$ to $x ^ p$.  $R$ is called *perfect* if this
endomorphism is an automorphism.  And under mild further conditions ($R$ an integral domain) this automorphism "lifts" <!-- check terminology -->

> to an automorphism of the field of fractions of $\mathbb{W}(R)$, which we will denote $\varphi$.  It's *Frobenius-semilinearity*, i.e. semilinearity with respect $\varphi$, that the number theorists

Reviewer 4: ```suggestion
to an automorphism of the field of fractions of $\mathbb{W}(R)$, which we will denote $\varphi$.  It's *Frobenius-semilinearity*, i.e. semilinearity with respect to $\varphi$, that the number theorists
```

> want to consider! 

Reviewer 2: If you want, you could say that by functoriality you get a Frobenius automorphism of `W(R)` for every `R`. Then, if `R` is an integral domain, so is `W(R)` and it makes sense to talk about the field of fractions. The automorphism will extend to the field of fractions because that's true for every automorphism of an integral domain.

Author: I'm not totally sure for what audience this will clarify things. My sense is that anyone who gets your two-sentence version will also get our (slightly vaguer) one-sentence version, or do you think that the extra detail would be helpful?

The fundamental result here is a classification theorem for

> Frobenius-semilinear automorphisms of finite-dimensional vector spaces over $K$, the field of fractions of $\mathbb{W}(k)$, for $k$ an algebraically closed field (and hence a perfect integral domain).  Such an object (a finite-dimensional 

Reviewer 2: Maybe introduce `K` as notation for the field of fractions in the preceding paragraph, when you also introduce `ϕ`?

Reviewer 5: I think it can't quite be done while remaining literally correct, because in the previous paragraph we discuss the fraction field of `𝕎 R`, but in this paragraph we discuss the fraction field of `𝕎 k`, for `k` some algebraically closed field.  So to remain literally correct one would have to refer to `K(R)`, `K(k)`, etc. which seems unwieldy.

![Rescale and reflect](/images/isocrystal-geo.png)

In general, the characterization of equivalence is the following: two $\sigma$-semilinear transformations $f_1(x)=a_1\sigma(x)$, $f_2(x)=a_2\sigma(x)$ of $F$ are equivalent if there exists a nonzero element $b\in F$ such that $a_1/a_2 = \sigma(b)/b$.  The set of elements of $F$ of the form $\sigma(b)/b$ forms a multiplicative subgroup of $F^*:=F\setminus\\{0\\}$. In the case of conjugation on $\mathbb{C}$ this subgroup is the unit circle.  And classifiying the equivalence classes of one-dimensional objects reduces to finding a family of canonical representatives of the cosets of this subgroup  -- in the case of $\mathbb{C}$, each coset of the unit circle action has a unique representative $r \in\mathbb{R}^+$.

# Frobenius-semilinear automorphisms of the fraction field of the Witt vectors

So, for the fraction field $K$ of the Witt vectors of an algebraically closed field $k$, what is this multiplicative subgroup

> $ \\{ \varphi(b) / b : b \in K^\* \\}$

Author: For whatever reason, this formats correctly with the `\\` and not with the `\`.

> $ \\{ \varphi(b) / b : b \in K^\* \\}$

Author: @faenuccio when I build the blog locally I see this with the `\\`:
![image](https://user-images.githubusercontent.com/4967469/182179839-e7a1dafa-e7d7-44e3-ad83-8bc73ffacd42.png)

And this with `\`:
![image](https://user-images.githubusercontent.com/4967469/182180076-57e075d7-e7d1-4497-957e-78cc12456afe.png)

of the nonzero elements 
$K^\*$ ?
Also, the cosets of this subgroup in $K^*$ are precisely in bijection with the isocrystals of dimension 1, so what is a complete set of representatives of the cosets?

At this point, we found a [2011 MathOverflow post](https://mathoverflow.net/questions/62468/about-frobenius-of-witt-vectors) from a drive-by pseudonym asking precisely this reformulated question.  ["Asker,"](https://mathoverflow.net/users/14572/asker) you gave no motivation for this rather specific and technical question ... we're still wondering whether you came up with it by the same route that we did!

So here is the answer.  The Witt vectors $\mathbb{W}(k)$ live inside their fraction field $K$ and there is a distinguished multiplicative subgroup $\mathbb{W}(k)^\times$ of $K^*$, the original units of $\mathbb{W}(k)$.  In the representation of $\mathbb{W}(k)$ as a sequence of elements of $k$, the units are [precisely](https://github.com/leanprover-community/mathlib/blob/lean-3.44.1/src/ring_theory/witt_vector/discrete_valuation_ring.lean#L58) the sequences whose first element is nonzero.

> The $p$-typical Witt vector $(0,1,0,\ldots)$ is referred to in $\mathbb{W}(k)$ as $p$ (it actually ends up being the sum of $p$ copies of $1=(1,0,0,\ldots)$, under the crazy Witt vector addition).  Multiplication (in the crazy Witt vector sense) by $p$ sends any Witt vector to a Witt vector with $0$ in the first position, and more generally multiplication by $p^m$ for any $m\in \mathbb{N}$ sends any Witt vector to a Witt vector with $m$ leading zeros. <!-- double-check how it works --> In fact, every nonzero Witt vector is [of the form $p^m w$](https://github.com/leanprover-community/mathlib/blob/lean-3.44.1/src/ring_theory/witt_vector/discrete_valuation_ring.lean#L117) for some $m\in\mathbb{N}$ and some $w\in\mathbb{W}(k)^\times$.  The statement for the fraction field $K$ is only slightly more complicated -- every nonzero element of $K$ is of the form $p^m w$ for some $m\in\mathbb{Z}$ and some Witt vector $w\in\mathbb{W}(k)^\times$. 

Reviewer 2: I personally always picture Witt vectors as trailing off to the left. Just like with `p`-adic "decimal expansions" of `p`-adic integers.

Reviewer 2: Note that
```
/-- The composition of Frobenius and Verschiebung is multiplication by `p`. -/
lemma frobenius_verschiebung (x : 𝕎 R) :
  frobenius (verschiebung x) = x * p :=
by { ghost_calc x, ghost_simp [mul_comm] }
```
tells you even more precisely what multiplication by `p` looks like, when `R` has characteristic `p`: you shift all the coefficients, introducing a `0` in the first position, and then you raise every coefficient to the power `p`. (Of course `0^p = 0`, so the first coeff is still `0`.)

Reviewer 3: Why do you use $^\ast$ to denote units of the field, and $^\times$ for units in the ring?

Reviewer 5: Implemented the two notational suggestions, thanks!  Regarding @jcommelin's comment about explaining what multiplication by `p` looks like, this seems like maybe too much detail for the blog post, but we added a link to that theorem on that sentence.  OK?

> The $p$-typical Witt vector $(0,1,0,\ldots)$ is referred to in $\mathbb{W}(k)$ as $p$ (it actually ends up being the sum of $p$ copies of $1=(1,0,0,\ldots)$, under the crazy Witt vector addition).  Multiplication (in the crazy Witt vector sense) by $p$ sends any Witt vector to a Witt vector with $0$ in the first position, and more generally multiplication by $p^m$ for any $m\in \mathbb{N}$ sends any Witt vector to a Witt vector with $m$ leading zeros. <!-- double-check how it works --> In fact, every nonzero Witt vector is [of the form $p^m w$](https://github.com/leanprover-community/mathlib/blob/lean-3.44.1/src/ring_theory/witt_vector/discrete_valuation_ring.lean#L117) for some $m\in\mathbb{N}$ and some $w\in\mathbb{W}(k)^\times$.  The statement for the fraction field $K$ is only slightly more complicated -- every nonzero element of $K$ is of the form $p^m w$ for some $m\in\mathbb{Z}$ and some Witt vector $w\in\mathbb{W}(k)^\times$. 

Reviewer 3: ```suggestion
The $p$-typical Witt vector $(0,1,0,\ldots)$ is referred to in $\mathbb{W}(k)$ as $p$ (it actually ends up being the sum of $p$ copies of $1=(1,0,0,\ldots)$, under the crazy Witt vector addition).  Multiplication (in the crazy Witt vector sense) by $p$ sends any Witt vector to a Witt vector with $0$ in the first position, and more generally multiplication by $p^m$ for any $m\in \mathbb{N}$ sends any Witt vector to a Witt vector with $m$ leading zeros. <!-- double-check how it works --> In fact, every nonzero Witt vector is [of the form](https://github.com/leanprover-community/mathlib/blob/lean-3.44.1/src/ring_theory/witt_vector/discrete_valuation_ring.lean#L117)  $p^m$ for some $m\in\mathbb{N}$ and some $w\in\mathbb{W}(k)^\times$.  The statement for the fraction field $K$ is only slightly more complicated -- every nonzero element of $K$ is of the form $p^m w$ for some $m\in\mathbb{Z}$ and some Witt vector $w\in\mathbb{W}(k)^\times$. 
```

Author: Don't have time to address all the comments right now, but just so it doesn't go unnoticed later, there's a `w` dropped in your suggestion! 

```suggestion
The $p$-typical Witt vector $(0,1,0,\ldots)$ is referred to in $\mathbb{W}(k)$ as $p$ (it actually ends up being the sum of $p$ copies of $1=(1,0,0,\ldots)$, under the crazy Witt vector addition).  Multiplication (in the crazy Witt vector sense) by $p$ sends any Witt vector to a Witt vector with $0$ in the first position, and more generally multiplication by $p^m$ for any $m\in \mathbb{N}$ sends any Witt vector to a Witt vector with $m$ leading zeros. <!-- double-check how it works --> In fact, every nonzero Witt vector is [of the form](https://github.com/leanprover-community/mathlib/blob/lean-3.44.1/src/ring_theory/witt_vector/discrete_valuation_ring.lean#L117)  $p^m w$ for some unique $m\in\mathbb{N}$ and some unique $w\in\mathbb{W}(k)^\times$.  The statement for the fraction field $K$ is only slightly more complicated -- every nonzero element of $K$ is of the form $p^m w$ for some $m\in\mathbb{Z}$ and some Witt vector $w\in\mathbb{W}(k)^\times$. 
```

Reviewer 3: Not sure I see it, but if you saw it once you will see it later again! ;-)

Reviewer 3: Right!

<a name="existence-lemma"></a>

> This subgroup $\mathbb{W}(k)^\times$ turns out to be precisely the things in $K$ which can be expressed as $\varphi(b)/b$ for some $b \in K$ (and it turns out that $b \in \mathbb{W}(k)$ suffices). Make a note of this statement. The rest of the math in this blog post will be devoted to proving this.

Reviewer 3: ```suggestion
This subgroup $\mathbb{W}(k)^\times$ turns out to be precisely the things in $K$ which can be expressed as $\varphi(b)/b$ for some $b \in K^\times$ (and it turns out that $b \in \mathbb{W}(k)\setminus {0}$ suffices). Make a note of this statement. The rest of the math in this blog post will be devoted to proving this.
```

[associated isocrystal](https://github.com/leanprover-community/mathlib/blob/lean-3.44.1/src/ring_theory/witt_vector/isocrystal.lean#L150):
$K$ itself (considered as a one-dimensional vector space over itself), equipped with the Frobenius-semilinear automorphism sending $x\in K$ to $p^m\varphi(x)$.  All one-dimensional isocrystals are equivalent to an isocrystal of this form. 

# Multiplication of Witt vectors

Given the structure of the statement we're working toward, it's not surprising that we'll need to dive into how multiplication on Witt vectors works. In this section we'll prove the
[following lemma](https://github.com/leanprover-community/mathlib/blob/lean-3.44.1/src/ring_theory/witt_vector/mul_coeff.lean#L275):

---
**Multiplication lemma.**
For every positive $n \in \mathbb{N}$, there is a function $f_n : k^{n} \times k^{n} \to k$ such that for every $x, y \in \mathbb{W}(k)$, the $n$-th coefficient of $x \cdot y$ is given by 

$$ x_{n}y_0^{p^{n}} + y_{n}x_0 ^{p^{n}} + f_n(x_0, \ldots, x_{n-1}, y_0, \ldots, y_{n-1}).  $$

> ---

Reviewer: This might work well as a blockquote (`>` at the start of each line)

Reviewer 5: I prefer it with our ad-hoc horizontal-line formatting, because with the indentation it's not visually clear that the displayed equation is part of the lemma.  OK?

And the proof-checking helped too. We mentioned above a wrinkle when proving the "multiplication lemma": some coefficients of the polynomial we were interested in vanish in characteristic $p$, so we had to work first with integer polynomials and later, after canceling a power of $p$, interpret them in the field $k$. This subtlety escaped us the first time around! 

![Oops...](/images/isocrystal-oops.png)

Back to our original motivation: the semilinear maps machinery worked great. As we had hoped, it sat invisibly in the background. Lean's extensible syntax was easy to use to set up notation <code style="white-space:nowrap;">M₁ ≃ᶠˡ[p, k] M₂</code> for the type of Frobenius-semilinear isomorphisms from `M₁` to `M₂`.

```lean 
notation M₁ ` ≃ᶠˡ[` p `,` k `] ` M₂ :=
  linear_equiv (witt_vector.fraction_ring.frobenius_ring_hom p k) M₁ M₂
``` 

From there, an isocrystal could be defined exactly as on paper: a $K$-module equipped with a Frobenius-semilinear automorphism.

```lean
class isocrystal (V : Type*) [add_comm_group V] 
  extends module K(p, k) V :=
( frob : V ≃ᶠˡ[p, k] V )
```

Check out the final statement of the classification theorem in mathlib 

> [here](https://github.com/leanprover-community/mathlib/blob/lean-3.44.1/src/ring_theory/witt_vector/isocrystal.lean#L163).

Reviewer: I think it would be wise to use a permalink here that includes a commit hash, just in case we ever patch old lean branches

---
author: 'Robert Y. Lewis, Heather Macbeth'
category: 'New in mathlib'
date: 2022-07-19 19:30:08 UTC+02:00
description: ''
has_math: true
link: ''
slug: classification-of-one-dimensional-isocrystals
tags: ''
title: Classification of one-dimensional isocrystals
type: text
---

> Last year, there was a [big mathlib refactor](link://filename/posts/semilinear-maps.md) to replace linear maps throughout the library with  *semilinear maps*,

Reviewer 6: In https://github.com/leanprover-community/blog/blob/isocrystals/posts/formalizing-isocrystals-in-lean.md the hyperlink is not showed.

Author: This syntax is specific to Nikola; it works when the blog is built!

It turns out that the standard use of semilinear maps in number theory is for *Frobenius-semilinearity*, semilinearity with
respect to the ring homomorphism of the fraction field of the $p$-typical Witt vectors over a perfect characteristic-$p$ integral domain which is induced by the Frobenius
automorphism of that domain.  Let's backtrack to catch everyone up...

<!-- TEASER_END -->

# Witt vectors and isocrystals

Back in 2020, one of us (Rob), together with Johan Commelin, [formalized the theory of Witt vectors](https://arxiv.org/abs/2010.02595) in Lean.  The $p$-typical Witt 
vectors $\mathbb{W}(R)$ over a ring $R$  are, concretely, sequences of elements of $R$, but equipped with a rather wild multiplicative and additive
structure (dependent on $p$) to make this set into a commutative ring.  The canonical example is the $p$-adic integers, which are the Witt vectors
of $\mathbb{Z}/p\mathbb{Z}$.

> A ring $R$ of characteristic $p$ has an endomorphism, constructed by sending each element $x$ to $x ^ p$.  $R$ is called *perfect* if this

Reviewer 6: ```suggestion
A ring $R$ of characteristic $p$ has an endomorphism, constructed by sending each element $x$ to $x ^ p$.  We say that $R$ is *perfect* if this
```
This is just to avoid ending a sentence with a symbol and starting the next one with another symbol.

endomorphism is an automorphism.  And under mild further conditions ($R$ an integral domain) this automorphism "lifts" 
to an automorphism of the field of fractions of $\mathbb{W}(R)$, which we will denote $\varphi$.  It's *Frobenius-semilinearity*, i.e. semilinearity with respect to $\varphi$, that the number theorists
want to consider![^1]

The fundamental result here is a classification theorem for

> Frobenius-semilinear automorphisms of finite-dimensional vector spaces over $K$, the field of fractions of $\mathbb{W}(k)$, for $k$ an algebraically closed field (and hence a perfect integral domain).  Such an object (a finite-dimensional 

Reviewer 6: ```suggestion
Frobenius-semilinear automorphisms of finite-dimensional vector spaces over $K$, the field of fractions of $\mathbb{W}(k)$, for $k$ an algebraically closed field of characteristic $p$ (and hence a perfect integral domain).  Such an object (a finite-dimensional 
```

Recall that a Witt vector $x$ over $k$ is given by a sequence $(\ldots, x_1, x_0)$ of elements of $k$. (Witt vectors are traditionally written as trailing to the left, to be suggestive of how we write numbers in decimal notation from largest place value to smallest.)
The $p$-typical Witt vector $(\ldots, 0,1,0)$ is referred to in $\mathbb{W}(k)$ as $p$ (it actually ends up being the sum of $p$ copies of $1=(\ldots, 0,0, 1)$, under the crazy Witt vector addition).  Multiplication (in the crazy Witt vector sense) by $p$ 
[sends any Witt vector](https://github.com/leanprover-community/mathlib/blob/25706131162465df8e6daecec75505f782fd428c/src/ring_theory/witt_vector/identities.lean#L37)

> to a Witt vector with $0$ in the first position, and more generally multiplication by $p^m$ for any $m\in \mathbb{N}$ sends any Witt vector to a Witt vector with $m$ leading zeros. In fact, every nonzero Witt vector is [of the form](https://github.com/leanprover-community/mathlib/blob/25706131162465df8e6daecec75505f782fd428c/src/ring_theory/witt_vector/discrete_valuation_ring.lean#L117)  $p^m w$ for some unique $m\in\mathbb{N}$ and some unique $w\in\mathbb{W}(k)^\times$.  The statement for the fraction field $K$ is only slightly more complicated -- every nonzero element of $K$ is of the form $p^m w$ for some $m\in\mathbb{Z}$ and some Witt vector $w\in\mathbb{W}(k)^\times$. 

Reviewer 5: @jcommelin Now that we wrote Witt vectors as trailing to the left, should we be changing the positional terminology?  For example, the reference here to 
> a Witt vector with $0$ in the first position

and a few paragraphs later to
> first $n$ coefficients

Is "first" still ok, or should we say "last coefficient", "lowest coefficient", ...?

<a name="existence-lemma"></a>
This subgroup $\mathbb{W}(k)^\times$ turns out to be precisely the things in $K$ which can be expressed as $\varphi(b)/b$ for some $b \in K^\times$ (and it turns out that $b \in \mathbb{W}(k)\setminus \\{0\\}$ suffices).[^4] Make a note of this statement. The rest of the math in this blog post will be devoted to proving this.

But to conclude the main discussion, our theorem is that every coset of the special subgroup has a representative of the form $p^m$ for some $m\in\mathbb{Z}$.  This integer $m$ is called the *slope* of the 
[associated isocrystal](https://github.com/leanprover-community/mathlib/blob/25706131162465df8e6daecec75505f782fd428c/src/ring_theory/witt_vector/isocrystal.lean#L150):

> $K$ itself (considered as a one-dimensional vector space over itself), equipped with the Frobenius-semilinear automorphism sending $x\in K$ to $p^m\varphi(x)$.  All one-dimensional isocrystals are equivalent to an isocrystal of this form. 

Reviewer 6: I find this sentence a little hard to follow. Can you be a little more explicit? Saying that you are building a "standard" isocrystal of slope $m$.

Author: @riccardobrasca I've reworded this paragraph:

> But to conclude the main discussion, our theorem is that every coset of the special subgroup has a representative of the form $p^m$ for some $m\in\mathbb{Z}$.
> We call $K$ (considered as a one-dimensional vector space over itself), equipped with the Frobenius-semilinear automorphism sending $x\in K$ to $p^m\varphi(x)$,
> the [standard isocrystal of slope $m$](https://github.com/leanprover-community/mathlib/blob/25706131162465df8e6daecec75505f782fd428c/src/ring_theory/witt_vector/isocrystal.lean#L150).
> All one-dimensional isocrystals are equivalent to one of these standard isocrystals. 

Does this sound better and accurate? cc @hrmacbeth

Reviewer 5: Looks good to me!
